Fix for Amasty filter "Availability" on product grid. By default this feature used old Magento inventory from table "cataloginventory_stock_item" and it's doesn't work with MSI.
Fix provided by Amasty with creativestyle adjustments
Issue appeared in ticket: IPET-1069

diff --git a/Plugin/Catalog/Ui/DataProvider/Product/ProductDataProvider.php b/Plugin/Catalog/Ui/DataProvider/Product/ProductDataProvider.php
index 5364735..d7bd803 100644
--- a/Plugin/Catalog/Ui/DataProvider/Product/ProductDataProvider.php
+++ b/Plugin/Catalog/Ui/DataProvider/Product/ProductDataProvider.php
@@ -183,11 +183,16 @@ class ProductDataProvider
         if (in_array(InlineEdit::AMASTY_AVAILABILITY, $visibleColumns)
             && !$subject->getCollection()->getFlag('amasty_instock_filter')
         ) {
-            $this->addInventoryColumn(
-                $subject->getCollection(),
-                InlineEdit::AMASTY_AVAILABILITY,
-                $this->availabilityColumn->getAvailableExpression()
+            // BEGIN PATCH
+            $this->addAvailabilityColumn(
+                $subject->getCollection()
             );
+            // $this->addInventoryColumn(
+            //     $subject->getCollection(),
+            //     InlineEdit::AMASTY_AVAILABILITY,
+            //     $this->availabilityColumn->getAvailableExpression()
+            // );
+            // END PATCH
         }

         if (in_array('amasty_backorders', $visibleColumns)) {
@@ -203,6 +208,20 @@ class ProductDataProvider
         }
     }

+    // BEGIN PATCH
+    protected function addAvailabilityColumn(Collection $collection): void
+    {
+        $collection->joinField(
+            'amasty_availability',
+            'inventory_source_item',
+            'IFNULL(status, 0)',
+            'sku=sku',
+            '{{table}}.status=1',
+            'left'
+        )->distinct(true);
+    }
+    // END PATCH
+
     private function addQtySoldColumn(Collection $collection): void
     {
         $collection->joinField(
diff --git a/Ui/DataProvider/Product/AddAvailabilityFilterToCollection.php b/Ui/DataProvider/Product/AddAvailabilityFilterToCollection.php
index e992259..7f2cebf 100644
--- a/Ui/DataProvider/Product/AddAvailabilityFilterToCollection.php
+++ b/Ui/DataProvider/Product/AddAvailabilityFilterToCollection.php
@@ -37,16 +37,27 @@ class AddAvailabilityFilterToCollection implements AddFilterToCollectionInterfac
             return;
         }

+        // BEGIN PATCH
         $collection->joinField(
             'amasty_availability',
-            'cataloginventory_stock_item',
-            $this->availabilityColumn->getAvailableExpression(),
-            'product_id=entity_id',
-            '{{table}}.stock_id=1',
-            'left'
-        );
-
-        $collection->getSelect()->where($this->availabilityColumn->getAvailableExpression() . '= ?', $condition['eq']);
+            'inventory_source_item',
+            'status',
+            'sku=sku',
+            '{{table}}.status=' . $condition['eq']
+        )->distinct(true);
+
+        // $collection->joinField(
+        //     'amasty_availability',
+        //     'cataloginventory_stock_item',
+        //     $this->availabilityColumn->getAvailableExpression(),
+        //     'product_id=entity_id',
+        //     '{{table}}.stock_id=1',
+        //     'left'
+        // );
+
+        // $collection->getSelect()->where($this->availabilityColumn->getAvailableExpression() . '= ?', $condition['eq']);
+        // END PATCH
+
         $collection->setFlag('amasty_instock_filter', 1);
     }
 }
